sunkit_image/granule.py

```diff
@@ -0,0 +1,391 @@
+       (e.g using cross_correlation() below).
+        ----------
+       Parameters:
+            data (numpy array): data to be clustered
+            llambda_axis (int): index for wavelength, -1 if scalar array.
+        ----------
+        Returns:
+            labels (numpy array): an array of labels, with 0 = granules,
+                                  2 = intergranules, 1 = in-between.
+            """
+
+    if llambda_axis not in [-1, 2]:
+        raise Exception('Wrong data shape. \
+        (either scalar or (x, y, llambda) )')
+    n_clusters = 3
+    n_init = 20
+    x_size = np.shape(data)[0]
+    y_size = np.shape(data)[1]
+    if llambda_axis == -1:
+        data_flat = np.reshape(data, (x_size * y_size, 1))
+        labels_flat = KMeans(n_clusters).fit(data_flat).labels_
+        labels = np.reshape(labels_flat, (x_size, y_size))
+    else:
+        llambda_size = np.shape(data)[llambda_axis]
+        data = np.reshape(data, (x_size * y_size, llambda_size))
+        labels = np.reshape(Kmeans(n_clusters, n_init).fit(data),
+                            (x_size, y_size))
+
+    # make intergranules 0, granules 1:
+    group0_mean = np.mean(data[labels == 0])
+    group1_mean = np.mean(data[labels == 1])
+    group2_mean = np.mean(data[labels == 2])
+
+    # intergranules
+    min_index = np.argmin([group0_mean,
+                           group1_mean,
+                           group2_mean])
+    segmented_map = np.ones(labels.shape)
+
+    segmented_map[[labels[:, :] == min_index][0]] -= 1
+
+    return segmented_map
+
+def cross_correlation(segment1, segment2):
+    """
+    Return -1 and print a message if the agreement between two
+    arrays is low, 0 otherwise. Designed to be used with segment
+    and segment_kmeans function.
+    ----------
+    Parameters:
+        segment1 (numpy array): 'main' array to compare the other input
+                                 array against.
+        segment2 (numpy array): 'other' array (i.e. data
+                                 segmented using kmeans).
+    ----------
+    Returns:
+        [label, confidence] (list): where
+        label is a label to summarize the confidence metric (int):
+            -1: if agreement is low (below 75%)
+            0: otherwise
+        confidence is the actual confidence metric (float):
+            float between 0 and 1 (0 if no agreement,
+                                   1 if completely agrees)
+    """
+
+    total_granules = np.count_nonzero(segment1 == 1)
+    total_intergranules = np.count_nonzero(segment1 == 0)
+
+    if total_granules == 0:
+        raise Exception('clustering problematic (no granules found)')
+
+    if total_intergranules == 0:
+        raise Exception('clustering problematic (no intergranules found)')
+
+    x_size = np.shape(segment1)[0]
+    y_size = np.shape(segment1)[1]
+
+    granule_agreement_count = 0
+    intergranule_agreement_count = 0
+    for i in range(x_size):

```

Comment:
Is there a way to do this overlap as an array multiplication, which would likely be much more efficient that looping through all the indices (16x10^6 in the case of a DKIST/VBI image)?

Reply:
Yes, there really should be. I can try to rewrite it, but I will first ask James why he did it this way.

Reply:
Is it possible for someone to test it on a file that sized?